### PR TITLE
MRG: Add cholesky solver for rERP

### LINKED ISSUE #3062: ENH: better solver for rERP, encoder models redux
Ridge and OLS as implemented in sklearn are much faster than the current solver, for some reason. (The reason is that obviously, something I hack together will always be worse than sklearn.)
They also are better at optionally dealing with sparse input.
I'd like to add them, and make them the go-to with the current one as a fallback.

(This is another incremental step of expanding rERP/linear_regression_raw into an encoder model.)

I'll go about implementing the first one soon.

Q: what's the most economic way of checking if a scipy coc sparse array is truly sparse, or alternatively, what's a good way of checking if it should be treated as sparse by the solver, or converted to dense and solved by a normal solver?

### PR DESCRIPTION
**Pull request by jona-sassenhagen**
closes #3062

### CODE REVIEW

File: mne/stats/regression.py [hunk 1]
```diff
@@ -228,8 +228,10 @@ def linear_regression_raw(raw, events, event_id=None, tmin=-.1, tmax=1,
     solver : str | function
         Either a function which takes as its inputs the sparse predictor
         matrix X and the observation matrix Y, and returns the coefficient
-        matrix b; or a string (for now, only 'pinv'), in which case the
-        solver used is `dot(scipy.linalg.pinv(dot(X.T, X)), dot(X.T, Y.T)).T`.
+        matrix b; or a string. If `pinv`,  the solver used is
+        `dot(scipy.linalg.pinv(dot(X.T, X)), dot(X.T, Y.T)).T`. If `cholesky`
```
Comment (Eric89GXL) [inline mne/stats/regression.py]: double backticks on code

File: mne/stats/regression.py [hunk 2]
```diff
@@ -246,6 +248,16 @@ def linear_regression_raw(raw, events, event_id=None, tmin=-.1, tmax=1,
...
+            try:
+                from sklearn.linear_model import Ridge
+                def solver(X, y):
+                    return Ridge(fit_intercept=False, copy_X=False, alpha=0.,
+                                 solver='cholesky').fit(X, y.T).coef_
+            except ImportError:
+                warn("Can't import from scikit-learn. Falling back to 'pinv'.")
+                solver == 'pinv'
+
         if solver == 'pinv':
```
Comment (Eric89GXL) [inline mne/stats/regression.py]: elif

Comment (Eric89GXL) [inline mne/stats/regression.py]: oh I see the defer to pinv above, never mind

File: mne/stats/regression.py [hunk 3]
```diff
@@ -246,6 +248,16 @@ def linear_regression_raw(raw, events, event_id=None, tmin=-.1, tmax=1,
...
 
     if isinstance(solver, string_types):
+        if solver == 'cholesky':
+            try:
+                from sklearn.linear_model import Ridge
+                def solver(X, y):
+                    return Ridge(fit_intercept=False, copy_X=False, alpha=0.,
+                                 solver='cholesky').fit(X, y.T).coef_
+            except ImportError:
+                warn("Can't import from scikit-learn. Falling back to 'pinv'.")
```
Comment (Eric89GXL) [inline mne/stats/regression.py]: actually I think this should be an error. If they ask for a method and it doesn't exist, error is more natural. (If we had an auto mode, or `cholesky` were the default, a warning might seem more appropriate.

Comment (Eric89GXL) [inline mne/stats/regression.py]: ...and instead of `try/except` we usually use a hopefully cleaner `sklearn` version check. If you look in `mne/cov.py` you should see some functions that simplify the checks.

File: mne/stats/regression.py [hunk 4]
```diff
@@ -246,6 +248,16 @@ def linear_regression_raw(raw, events, event_id=None, tmin=-.1, tmax=1,
     """
 
     if isinstance(solver, string_types):
+        if solver == 'cholesky':
+            try:
+                from sklearn.linear_model import Ridge
+                def solver(X, y):
+                    return Ridge(fit_intercept=False, copy_X=False, alpha=0.,
+                                 solver='cholesky').fit(X, y.T).coef_
```
Comment (agramfort) [inline mne/stats/regression.py]: you don't need sklearn for this. What it does is just:

```
linalg.solve(np.dot(X.T, X) , np.dot(X.T, y), sym_pos=True, overwrite_a=True)
```

Comment (jona-sassenhagen) [inline mne/stats/regression.py]: I had originally planned on just doing that, but sklearn also uses safe_sparse_dot (x is sparse here) and it turned out to be more complicated importing that and writing the solver than just importing ridge. Or is safe_sparse_dot not important?






On Sun, Mar 27, 2016 at 12:27 PM -0700, "Alexandre Gramfort" <notifications@github.com> wrote:












In mne/stats/regression.py:
> @@ -246,6 +248,16 @@ def linear_regression_raw(raw, events, event_id=None, tmin=-.1, tmax=1,
>      """
>  
>      if isinstance(solver, string_types):
> +        if solver == 'cholesky':
> +            try:
> +                from sklearn.linear_model import Ridge
> +                def solver(X, y):
> +                    return Ridge(fit_intercept=False, copy_X=False, alpha=0.,
> +                                 solver='cholesky').fit(X, y.T).coef_



you don't need sklearn for this. What it does is just:

linalg.solve(np.dot(X.T, X) , np.dot(X.T, y), sym_pos=True, overwrite_a=True)

Comment (agramfort) [inline mne/stats/regression.py]: if you know that X is sparse just use sparse dot. it's an overkill to use
sklearn for this.

Comment (jona-sassenhagen) [inline mne/stats/regression.py]: Ok!

Comment (agramfort) [inline mne/stats/regression.py]: and if cholesky is always faster just replace the current solver without
adding the parameter :)

thx

Comment (jona-sassenhagen) [inline mne/stats/regression.py]: The parameter is already there actually - it's important, because it allows to plug in default solvers. But in principle, 'pinv' could be replaced with 'cholesky' - in my personal tests, it's always much faster - but I haven't done a systematic check.

**Comment by jona-sassenhagen** [comment]
I'm getting a 5x speedup on the sample data with this.

**Comment by agramfort** [comment]
why note removing the pinv option then?

**Comment by jona-sassenhagen** [comment]
Done

**Comment by jona-sassenhagen** [comment]
I did some checks on the moderately dense case and cholesky is still faster.

**Comment by agramfort** [comment]
no test update?

please the PR to MRG when ready

**Comment by jona-sassenhagen** [comment]
Tests don't need to change because we expect the same output, right? It's just faster.

**Comment by agramfort** [comment]
ok fair enough. Can you update what's new? thx

**Comment by jona-sassenhagen** [comment]
All done.

**Comment by jona-sassenhagen** [comment]
When this one is finished: In initial experiments, I'm getting another 5-fold speedup over this when offloading this to a GPU. Can we have an additional CUDA-based solver, using pycuda and skcuda?

**Comment by Eric89GXL** [comment]
> Can we have an additional CUDA-based solver, using pycuda and skcuda?

In my little experiments I haven't gotten much advantage from offloading linear algebra to CUDA, since memory transfer overhead limited things. But if you've found an application that can leverage GPU to get those gains, `n_jobs=1 | 'cuda'` sounds good to me.

File: mne/stats/regression.py [hunk 5]
```diff
@@ -246,13 +246,12 @@ def linear_regression_raw(raw, events, event_id=None, tmin=-.1, tmax=1,
     """
 
     if isinstance(solver, string_types):
-        if solver == 'pinv':
-            fast_dot = _get_fast_dot()
+        if solver == 'cholesky':
+            def solver(X, y):
+                a = (X.T * X).toarray()
```
Comment (Eric89GXL) [inline mne/stats/regression.py]: I prefer `X.T.dot(X)`, since it makes clear it's a dot product. Otherwise I have to try and assume/figure it outfrom the code. 99.9% (or 100%) of other times in our code `*` means broadcast-multiply, not dot.

Comment (jona-sassenhagen) [inline mne/stats/regression.py]: Tried that, it didn't work. I think sparse matrices don't have the same methods as arrays (something about the matrix object being generally weird). So we could first convert it, but I think it's much more efficient to first do the dot product and then covert?

Comment (Eric89GXL) [inline mne/stats/regression.py]: This is fine then, but please add a comment saying X is a matrix object
therefore * means dot

**Comment by jona-sassenhagen** [comment]
Done!

File: doc/whats_new.rst [hunk 6]
```diff
@@ -127,6 +127,9 @@ API
 
     - EOG, ECG and EMG channels are now plotted by default (if present in data) when using :func:`mne.viz.plot_evoked` by `Marijn van Vliet`_
 
+
+    - Replace pseudoinverse-based solver with much faster Cholesky solver in :func:`mne.stats.regression.linear_regression_raw`, by `Jona Sassenhagen`_.
```
Comment (Eric89GXL) [inline doc/whats_new.rst]: FYI `linear_regression_raw` actually lives more readily accesible in the `mne.stats` namespace

File: mne/stats/regression.py [hunk 7]
```diff
@@ -228,8 +228,8 @@ def linear_regression_raw(raw, events, event_id=None, tmin=-.1, tmax=1,
     solver : str | function
         Either a function which takes as its inputs the sparse predictor
         matrix X and the observation matrix Y, and returns the coefficient
-        matrix b; or a string (for now, only 'pinv'), in which case the
-        solver used is `dot(scipy.linalg.pinv(dot(X.T, X)), dot(X.T, Y.T)).T`.
+        matrix b; or a string. If str, must be `cholesky`, in which case the
```
Comment (Eric89GXL) [inline mne/stats/regression.py]: double-backticks on code (cholesky), or use quotes if you want. I'd probably do quotes inside double-backticks

**Comment by Eric89GXL** [comment]
I'll make the final minor changes myself when merging once the CIs come back happy

File: doc/whats_new.rst (hunk 6 above)
Comment (jona-sassenhagen) [inline doc/whats_new.rst]: I actually didn't know that.

**Comment by jona-sassenhagen** [comment]
Thanks Eric, and Alex. Wheeeeew more speed!

**Comment by Eric89GXL** [comment]
Thanks @jona-sassenhagen
